chore: Lint and format only staged changes

AlexVascon opened the pull request:
## **User description**
## Description

This commit adds the package lint-staged. This package makes sure only staged changes are scanned for linting and formating instead of every file. I think when lint-staged is calling pnpm run lint it passes some file arguments. Since pnpm run lint calls turbo run lint this argument is getting passed directly resulting in an error as extra arguments get added to the tubro command. to prevent this we add '--' to signigy not to pass on further arguments.

I based this from a comment made by github user mehulkar whos explanation and solution helped me solve this problem. 
link [here](https://github.com/vercel/turbo/issues/3364#issuecomment-1396687996)

_Give a summary of the change that you have made_ <br />

Changes are made to 2 files. 

package.json root
.husky/pre-commit

In the package.json a section is included for lint-staged. Within 2 commands are set so only lint and formating are run on staged commits and not the whole file. These commands where moved from the husky pre-commit file and replaced with  pnpm run lint-staged. This calls the code under lint-staged in the package.json. 

Fixes #195 []
issue 195 
link [here](https://github.com/keyshade-xyz/keyshade/issues/195)

This commit fixes the issue where all files got scanned for linting and formating which is very slow and it's better that only the staged files get scanned as they contain changes. 

## Dependencies

[lint-staged](https://www.npmjs.com/package/lint-staged/v/12.3.2)

_Mention any dependencies/packages used_

lint-staged

## Future Improvements

_Mention any improvements to be done in future related to any file/feature_

## Mentions

_Mention and tag the people_

## Screenshots of relevant screens
<img width="768" alt="Screenshot 2024-05-07 at 16 21 09" src="https://github.com/keyshade-xyz/keyshade/assets/80917151/94bf9511-b527-46ce-abff-5b7f698f0e6e">

<img width="898" alt="Screenshot 2024-05-07 at 16 22 12" src="https://github.com/keyshade-xyz/keyshade/assets/80917151/73452bb4-9b75-4fa8-9ef4-6df610a3b9e4">


This screenshot below is what it looks like if you dont use lint-staged. I edited the pre-commit file and changed pnpm lint-staged to pnpm lint.

<img width="932" alt="Screenshot 2024-05-07 at 16 58 52" src="https://github.com/keyshade-xyz/keyshade/assets/80917151/2c5f5504-4b36-4734-9ba8-e0745363fc9e">


_Add screenshots of relevant screens_

## Developer's checklist

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-check on my work

**If changes are made in the code:**

- [ x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x ] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [ x] There are no UI/UX issues


### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [x ] I have made the necessary updates to the documentation, or no documentation changes are required.


___

## **Type**
enhancement, bug_fix


___

## **Description**
- Integrated `lint-staged` in `package.json` to ensure only staged files are linted and formatted, improving performance.
- Updated the Husky pre-commit hook to use `lint-staged`, streamlining the pre-commit checks.
- Added `lint-staged` to `devDependencies` to manage the new dependency.


___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Integrate lint-staged for Efficient Linting and Formatting</code></dd></summary>
<hr>

package.json
<li>Added a new <code>lint-staged</code> configuration to handle linting and formatting <br>of staged files.<br> <li> Updated <code>devDependencies</code> to include <code>lint-staged</code> version <code>^15.2.2</code>.


</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/197/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519">+10/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pre-commit</strong><dd><code>Update Husky Pre-commit Hook to Use lint-staged</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.husky/pre-commit
<li>Replaced direct lint and format commands with <code>npx lint-staged</code> to <br>utilize the new lint-staged configuration.<br> <li> Retained existing <code>pnpm test:api</code> command.


</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/197/files#diff-d2bc4bbf14eadc292d84e0ef5f7a93115c23557f533809fc80b896934291529d">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

rajdip-b commented on .husky/pre-commit:
```diff
@@ -1,4 +1,4 @@
 
-pnpm lint && pnpm format && pnpm test:api
\ No newline at end of file
+npx lint-staged && pnpm test:api
```
Since we are using pnpm, i believe `pnpx` would be a better thing to use in here.

kriptonian1 replied:
Why have we removed pnpm format from here

kriptonian1 replied:
We can use task concurrency ig @rajdip-b  
https://www.npmjs.com/package/lint-staged#task-concurrency

rajdip-b commented:
@kriptonian1 Can you share your reviews on this?

kriptonian1 commented:
@AlexVascon  Also do this a check https://www.npmjs.com/package/lint-staged#how-to-use-lint-staged-in-a-multi-package-monorepo
since it's a monorepo, we should follow the format they mentioned in the doc

kriptonian1 reviewed (changes requested):
I believe that once this review has been resolved, we will be prepared to proceed.

rajdip-b commented:
hey bro @AlexVascon, could you please update us on this?

AlexVascon commented:
Hi sorry I do apologise. I was dealing with a medical issue. I will begin addressing your concerns right now

rajdip-b commented:
> Hi sorry I do apologise. I was dealing with a medical issue. I will begin addressing your concerns right now

Hey man. I hope you are okay! Incase you need more time, please feel free to. Health is important!

AlexVascon commented:
> @AlexVascon Also do this a check https://www.npmjs.com/package/lint-staged#how-to-use-lint-staged-in-a-multi-package-monorepo since it's a monorepo, we should follow the format they mentioned in the doc

Please correct me if I am wrong but my understanding is you guys are using turbo which acts as a kind of central hub to execute commands across all packages in the monorepo. By defining linting at the root level and invoking it using Turbo mode (turbo run lint), we have centralized the linting configuration for your entire monorepo. This means that when you run linting commands from the root, they apply to all packages in the monorepo. (I had some help from chatGPT to help me understand)

This makes sense since initially in the pre-commit file you ran pnpm lint which in the root package.json runs the script "lint": "turbo run lint".

rajdip-b commented:
Yes, your understanding is correct.

AlexVascon commented on package.json:
```diff
@@ -121,11 +121,21 @@
     "sourcemaps:api": "turbo run sourcemaps --filter=api"
   },
+  "lint-staged": {
+    "*.{js,ts,tsx}": [
```
This is true. I have added more extensions that I found your project uses. Let me know if you would like any further changes. I could not add .json files for the linting as you do not have any configuration for it. You use various eslint-plugins but not eslint-plugin-json. Would you like me to add it? otherwise I can only run the formatter which is why I added a seperate script for it in the lint-stage

AlexVascon commented:
Ok based on feedback I have made some changes. 
- I changed the pre-commit script to use pnpm instead of npx. That was an oversite on my part. 
- I also added all the file extensions I could find to add to the stage-lint script. Please let me know if you would like anything changed. 
- I could not add .json files for linting as you do not have any file extensions for it. You use various eslint-plugins but not eslint-plugin-json. Would you like me to add it? Otherwise I had to create a seperate script so only the formating is run for .json files
- I also took some pictures which I will add to the PR description. These pictures represent the during and after the linting is run

rajdip-b commented:
Hey @AlexVascon! Sorry for the late reply.

Thanks for putting in all that effort!
- No we won't need `.json` linting right yet because we aren't using any json files in particular
- It would be real nice if you can upload those screenshots so that we know what to expect!
- Also, can you resolve the merge conflict? Try pulling the `develop` branch into your local `develop` branch, and **rebase** your current branch with your local `develop`. If you find merge conflicts with the lock file, try deleting your lockfile and running `pnpm i`. This should get it fixed. Then, make a push!

rajdip-b commented:
Hey @AlexVascon ! Any updates on this?

kriptonian1 commented:
Hey @AlexVascon what's the upate

kriptonian1 commented:
@AlexVascon are you still working on this issue

rajdip-b commented:
Closing because of inactivity
